docs/rfcs/751-write-empty-file-behavior.md

```diff
@@ -0,0 +1,51 @@
+- Author:  Abyss-w <mad.hatter@foxmail.com>
+- Start Date: 2021-09-10
+- RFC PR: [beyondstorage/go-storage#751](https://github.com/beyondstorage/go-storage/pull/751)
+- Tracking Issue: [beyondstorage/go-storage#0](https://github.com/beyondstorage/go-storage/issues/0)
+
+# GSP-751: Write Empty File Behavior
+
+## Background
+
+In our definition, the `Write` API will upload a file to the target object. We do not do anything about uploading empty files. However, some services do not support uploading empty files. As [proposed](https://forum.beyondstorage.io/t/topic/204) in our community, for s3, when we upload an empty file, the service will hang permanently. Here is an example of s3:
+
+```go
+_, err = fs.s.Write(path,nil, 0)
+if err != nil {
+    fs.logger.Error("write", zap.String("path", path), zap.Error(err))
+    return nil, nil, err
+}
+```
+
+If we want to upload an empty file, we need to do so:
+
+```go
+_, err = fs.s.Write(path, bytes.NewReader([]byte{}), 0)
+if err != nil {
+    fs.logger.Error("write", zap.String("path", path), zap.Error(err))
+    return nil, nil, err
+}
+```
+
+This is not convenient for the users.
+
+## Proposal
+
+I propose to allow the user to pass in an empty file when calling `Write`. 
+
+- For services that do not support uploading empty files, we should check if the `reader` is `nil`. If it is `nil`, we need to create a `reader` before calling the API. Like s3, kodo, etc.

```

Comment:
Maybe we should describe it differently instead of "empty file" as services like s3, kodo support upload or empty file or object with size 0, but `io.Reader` is not allowed to be `nil` for the `Write` APIs?

Reply:
Got it!